google/gax/__init__.py

```diff
@@ -482,3 +489,160 @@ def __next__(self):
+        return 'ResultError({}, with code {})'.format(msg, self.code)
+
+
+class OperationFuture(object):
+    """A Future which polls a service for completion via OperationsApi."""
+
+    def __init__(self, operation, client, result_type, metadata_type,
+                 call_options=None):
+        """Constructor.
+
+        Args:
+            operation (google.longrunning.Operation): the initial long-running
+                operation.
+            client (google.cloud.gapic.longrunning.operations_api.OperationsApi):
+                the client used to manage the long-running operation with an API
+                service.
+            result_type (type): the class type to be unpacked from the result.
+            metadata_type (type): the class type to be unpacked from the
+                metadata.
+            call_options (google.gax.CallOptions, optional): the call options
+                that are used when reloading the operation.
+        """
+        self._first_operation = operation
+        self._last_operation = operation
+        self._client = client
+        self._result_type = result_type
+        self._metadata_type = metadata_type
+        self._call_options = call_options
+        self._cancelled = False
+
+    def cancel(self):
+        """If last Operation's value of `done` is true, returns false;
+           otherwise, issues OperationsApi.cancel_operation and returns true.
+        """
+        if not self._last_operation.done:
+            self._client.cancel_operation(self._last_operation.name)
+            self._cancelled = True
+
+        return self._cancelled
+
+    def result(self, timeout=None):
+        """Enters polling loop on OperationsApi.get_operation, and once Operation.done
+           is true, then returns Operation.response if successful or throws
+           ResultError if not successful.
+
+           This method will wait up to timeout seconds. If the call hasn't
+           completed in timeout seconds, then a concurrent.futures.TimeoutError
+           will be raised. timeout can be an int or float. If timeout is not
+           specified or None, there is no limit to the wait time.
+        """
+        if not self._poll(timeout).HasField('response'):
+            raise ResultError.from_operation_error(self._last_operation.error)
+
+        return _from_any(self._result_type, self._last_operation.response)
+
+    def exception(self, timeout=None):
+        """Similar to result(), except returns the exception if any."""
+        if self._poll(timeout).HasField('error'):
+            return self._last_operation.error
+
+        return None
+
+    def cancelled(self):
+        """Return True if the call was successfully cancelled."""
+        return self._cancelled
+
+    def done(self):
+        """Issues OperationsApi.get_operation and returns value of Operation.done."""
+        return self._get_operation().done
+
+    def add_done_callback(self, done_clbk):
```

Comment:
A few thoughts on this:

First, while it is a bit of a nitpick, I would ask that the positional argument be spelled `fn` to match `c.f.Future` and `asyncio`. It is possible to pass positional arguments using their keyword names, and many people do.

Second, I am having a debate with myself over the thread here. Part of me thinks that this method should _return_ the `Thread` object (so that the programmer would have the option of calling `thread.join()` later). On the other hand, I do not particularly want to be tied to this exact implementation; we might want to change this to use gevent or something. Since `c.f.Future` and `asyncio` do not expect this to return, probably we should leave it.

Finally, it is worth noting that the `threading` module is relatively poorly implemented in Python and is generally disfavored (with [`multiprocess`][1] being preferred). If this is something you can change quickly, it would probably be preferable. That said, I also think it is not the end of the world if that is time-consuming, especially since @bjwatson has already fingered this spot for a potential improvement as we discover how it is actually used.

  [1]: https://docs.python.org/2/library/multiprocessing.html#module-multiprocessing